global_controller/global_controller.py

```diff
@@ -0,0 +1,181 @@
+from common.configurable_model import ConfigurableModel
+from bit_vector import BitVector
+from enum import Enum
+import magma as m
+import fault
+
+
+class GCRegAddr(Enum):
+    TST_ADDR = 0
+    STALL_ADDR = 1
+    CLK_SEL_ADDR = 2
+    RW_DELAY_SEL_ADDR = 3
+    CLK_SWITCH_DELAY_SEL_ADDR = 4
+
+
+class GCOp(Enum):
+    NOP = 0
+    CONFIG_WRITE = 1
+    CONFIG_READ = 2
+    WRITE_A050 = 4
+    WRITE_TST = 5
+    READ_TST = 6
+    GLOBAL_RESET = 7
+    WRITE_STALL = 8
+    READ_STALL = 9
+    ADVANCE_CLK = 10
+    READ_CLK_DOMAIN = 11
+    SWITCH_CLK = 12
+    WRITE_RW_DELAY_SEL = 13
+    READ_RW_DELAY_SEL = 14
+    WRITE_CLK_SWITCH_DELAY_SEL = 15
+    READ_CLK_SWITCH_DELAY_SEL = 16
+
+
+def gen_global_controller(config_data_width: int,
+                          config_addr_width: int,
+                          config_op_width: int):
+
+    class _GlobalController(ConfigurableModel(32, 32)):
+        def __init__(self):
+            super().__init__()
+            self.num_stall_domains = 4
+            self.reset()
+
+        def reset(self):
+            self.TST = [BitVector(0, config_data_width)]
+            self.stall = [BitVector(0, self.num_stall_domains)]
+            self.clk_sel = [BitVector(0, 1)]
+            self.rw_delay_sel = [BitVector(2, config_data_width)]
+            self.clk_switch_delay_sel = [BitVector(0, 1)]
+
+            self.reset_out = [0]
+            self.config_addr_out = [BitVector(0, config_addr_width)]
+            self.config_data_out = [BitVector(0, config_data_width)]
+            self.config_data_in = fault.UnknownValue
+            self.read = [0]
+            self.write = [0]
+            self.config_data_to_jtag = [BitVector(0, config_data_width)]
+
+        def CONFIG_READ(self, addr):
```

Comment:
nit: `config_read()`

Reply:
oops... I did a replace all to fix the other case errors, but I forgot the functions had the same names as the ops...